pennylane/math/single_dispatch.py

```diff
@@ -0,0 +1,327 @@
+
+
+ar.register_function("autograd", "scatter_element_add", _scatter_element_add_autograd)
+
+
+def _take_autograd(tensor, indices, axis=None):
+    indices = _i("qml").numpy.asarray(indices)
+
+    if axis is None:
+        return tensor.flatten()[indices]
+
+    fancy_indices = [slice(None)] * axis + [indices]
+    return tensor[tuple(fancy_indices)]
+
+
+ar.register_function("autograd", "take", _take_autograd)
+
+
+# -------------------------------- TensorFlow --------------------------------- #
+
+
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "angle"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "arcsin"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "arccos"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "arctan"] = "tensorflow.math"
+ar.autoray._SUBMODULE_ALIASES["tensorflow", "diag"] = "tensorflow.linalg"
+
+
+ar.autoray._FUNC_ALIASES["tensorflow", "arcsin"] = "asin"
+ar.autoray._FUNC_ALIASES["tensorflow", "arccos"] = "acos"
+ar.autoray._FUNC_ALIASES["tensorflow", "arctan"] = "atan"
+ar.autoray._FUNC_ALIASES["tensorflow", "diag"] = "diag"
+
+
+ar.register_function("tensorflow", "asarray", lambda x: _i("tf").convert_to_tensor(x))
+ar.register_function("tensorflow", "flatten", lambda x: _i("tf").reshape(x, [-1]))
+ar.register_function("tensorflow", "shape", lambda x: tuple(x.shape))
+ar.register_function(
+    "tensorflow",
+    "sqrt",
+    lambda x: _i("tf").math.sqrt(
+        _i("tf").cast(x, "float64") if x.dtype.name in ("int64", "int32") else x
+    ),
+)
+
+
+def _take_tf(tensor, indices, axis=None):
+    tf = _i("tf")
+
+    if isinstance(indices, numbers.Number):
+        indices = [indices]
+
+    indices = tf.convert_to_tensor(indices)
+
+    if np.any(indices < 0):
+        # Unlike NumPy, TensorFlow doesn't support negative indices.
+        dim_length = tf.size(tensor).numpy() if axis is None else tf.shape(tensor)[axis]
+        indices = tf.where(indices >= 0, indices, indices + dim_length)
+
+    if axis is None:
+        # Unlike NumPy, if axis=None TensorFlow defaults to the first
+        # dimension rather than flattening the array.
+        data = tf.reshape(tensor, [-1])
+        return tf.gather(data, indices)
+
+    return tf.gather(tensor, indices, axis=axis)
+
+
+ar.register_function("tensorflow", "take", _take_tf)
+
+
+def _coerce_types_tf(tensors):
+    tf = _i("tf")
+    tensors = [tf.convert_to_tensor(t) for t in tensors]
+    dtypes = {i.dtype for i in tensors}
+
+    if len(dtypes) == 1:
+        return tensors
+
+    complex_type = dtypes.intersection({tf.complex64, tf.complex128})
+    float_type = dtypes.intersection({tf.float16, tf.float32, tf.float64})
+    int_type = dtypes.intersection({tf.int8, tf.int16, tf.int32, tf.int64})
+
+    cast_type = complex_type or float_type or int_type
+    cast_type = list(cast_type)[-1]
+
+    return [tf.cast(t, cast_type) for t in tensors]
+
+
+ar.register_function("tensorflow", "coerce", _coerce_types_tf)

```

Comment:
@antalszava here, `_coerce_types_tf` is being registered as the TensorFlow version of the `np.coerce` single dispatch function. So `np.coerce`, called with a tensorflow argument, will then be dispatched to `_coerce_types_tf`.

Similar registrations are done for torch (below), and numpy and autograd (above)